File: 1986-largest-color-value-in-a-directed-graph/largest-color-value-in-a-directed-graph.java
class Solution {
    public int largestPathValue(String colors, int[][] edges) {
        ArrayList<Integer> [] edgeList = new ArrayList[colors.length()];
        for(int i = 0;i<colors.length();i++){
            edgeList[i] = new ArrayList<>();
        }   
        for(int i = 0;i<edges.length;i++){
            edgeList[edges[i][0]].add(edges[i][1]);
        }
        int max = 0;
        int [] visited = new int[colors.length()];
        Arrays.fill(visited,0);
        int [] path = new int[colors.length()];
        Arrays.fill(path,0);
        int [][] memory = new int[colors.length()][26];
        for(int i = 0;i<colors.length();i++){
            if(dfs(i,memory,visited,path,edgeList,colors) == -1){
                return -1;
            }
        }
        for(int i = 0;i<colors.length();i++){
            for(int j = 0;j<26;j++){
                max = Math.max(max,memory[i][j]);
            }
        }
        return max;
    }
    public int dfs(int node,int [][] memory,int [] visited,int [] path,ArrayList<Integer> [] edgeList,String colors){
        visited[node] = 1;
        path[node] = 1;
        int [] colorCount = new int[26];
        for(int nextNode : edgeList[node]){
            if(path[nextNode] == 1){
                return -1;
            }
            else{
                if(visited[nextNode] == 0){
                    if(dfs(nextNode,memory,visited,path,edgeList,colors) == -1){
                        return -1;
                    }
                }
            }
            for(int i = 0;i<26;i++){
                colorCount[i] = Math.max(colorCount[i],memory[nextNode][i]);
            }
        }
        colorCount[colors.charAt(node) - 'a']+=1;
        memory[node] = colorCount;
        path[node] = 0;
        return 0;
    }
}

File: 1986-largest-color-value-in-a-directed-graph/largest-color-value-in-a-directed-graph.py
class Solution:
    def largestPathValue(self, colors: str, edges: List[List[int]]) -> int:
        n = len(colors)
        adjList = [[] for _ in range(n)]
        for edge in edges:
            adjList[edge[0]].append(edge[1])
        dp = {}

        def DFS(node,visited, path):
            if len(adjList[node]) == 0:
                data = [0 for _ in range(26)]
                data[ord(colors[node]) - ord('a')]+=1
                dp[node] = data
                return data
            data = [0 for _ in range(26)]
            for adjNode in adjList[node]:
                if visited[adjNode] == -1:
                    visited[adjNode] = 1
                    path.add(adjNode)
                    returnedData = DFS(adjNode,visited,path)
                    if returnedData == -1:
                        return -1
                    path.remove(adjNode)
                else:
                    if adjNode in path:
                        return -1
                    else:
                        returnedData = dp[adjNode]
                data = [max(di,ri) for di,ri in zip(data,returnedData)]
            data[ord(colors[node]) - ord('a')]+=1
            dp[node] = data
            return data
        visited = [-1 for _ in range(n)]
        path = set()
        result = 0
        for i in range(n):
            if visited[i] == -1:
                visited[i] = 1
                path.add(i)
                returnedData = DFS(i,visited,path)
                if returnedData == -1:
                    return -1
                result = max(result, max(returnedData))
                path.remove(i)
        return result
